### ISSUE TITLE
Cannot replicate the CImg API example

### ISSUE THREAD

**Issue by MichelV2** [dahtah/imager#7]
Hello,

Great package! I am just struggling to replicate the following example given in the documentation:
`library(imager)
library(Rcpp)

#Some C++ code as a character string
#The function takes an image, applies blur and erode operations, and returns the result
foo.inline <- "
NumericVector foo(NumericVector inp)
{
    CImg<double> img = as<CImg<double> >(inp);
    img.blur(3).erode(4);
    return wrap(img);
} "

#Compile the function. The compiler may issue a number of warnings which you can all ignore
cppFunction(foo.inline,depends="imager")`

The last line gives me the following error:

> g++ -m64 -I"*/R-32~1.2/include" -DNDEBUG     -I*/R-3.2.2/library/Rcpp/include" -I"*/R-3.2.2/library/imager/include" -I"*/R-3.2.2/library/Rcpp/include" -I"*/RCompile/r-compiling/local/local320/include"     -O2 -Wall  -mtune=core2 -c file275023be36bb.cpp -o file275023be36bb.o
> In file included from *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/objbase.h:154:0,
>                  from *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/ole2.h:16,
>                  from *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/shlobj.h:86,
>                  from */R-3.2.2/library/imager/include/CImg.h:171,
>                  from */R-3.2.2/library/imager/include/imager.h:21,
>                  from file275023be36bb.cpp:3:
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/objidl.h:598:52: error: macro "Realloc" requires 3 arguments, but only 2 given
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/objidl.h:598:56: error: ISO C++ forbids initialization of member 'Realloc' [-fpermissive]
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/objidl.h:598:56: error: making 'Realloc' static [-fpermissive]
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/objidl.h:598:56: error: invalid in-class initialization of static data member of non-integral type 'void*'
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/objidl.h:599:25: error: expected identifier before '(' token
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/objidl.h:599:25: error: 'parameter' declared as function returning a function
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/objidl.h:599:25: error: expected ')' before ',' token
> In file included from **\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/msxml.h:209:0,
>                  from *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/urlmon.h:219,
>                  from *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/objbase.h:348,
>                  from *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/ole2.h:16,
>                  from *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/shlobj.h:86,
>                  from */R-3.2.2/library/imager/include/CImg.h:171,
>                  from */R-3.2.2/library/imager/include/imager.h:21,
>                  from file275023be36bb.cpp:3:
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/oaidl.h:1890:28: error: expected identifier before '(' token
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/oaidl.h:1890:28: error: 'parameter' declared as function returning a function
> *\rbuildtools\3.3\gcc-4.6.3\bin\../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/include/oaidl.h:1890:28: error: expected ')' before ',' token
> In file included from **/R-3.2.2/library/imager/include/imager.h:21:0,
>                  from file275023be36bb.cpp:3:
> */R-3.2.2/library/imager/include/CImg.h: In function 'const char* cimg_library::cimg::strbuffersize(UINT_PTR)':
> */R-3.2.2/library/imager/include/CImg.h:56410:82: warning: format '%lu' expects argument of type 'long unsigned int', but argument 4 has type 'UINT_PTR {aka long long unsigned int}' [-Wformat]
> In file included from **/R-3.2.2/library/imager/include/imager.h:23:0,
>                  from file275023be36bb.cpp:3:
> */R-3.2.2/library/imager/include/wrappers.h: In function 'SEXPREC* Rcpp::wrap(const T&) [with T = cimg_library::CImgList<double>, SEXP = SEXPREC*]':
> */R-3.2.2/library/imager/include/wrappers.h:39:35: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
> make: *** [file275023be36bb.o] Error 1
> Warning message:
> running command 'make -f "*/R-32~1.2/etc/x64/Makeconf" -f "*/R-32~1.2/share/make/winshlib.mk" SHLIB_LDFLAGS='$(SHLIB_CXXLDFLAGS)' SHLIB_LD='$(SHLIB_CXXLD)' SHLIB="sourceCpp_54.dll" WIN=64 TCLBIN=64 OBJECTS="file275023be36bb.o"' had status 2 
> Error in sourceCpp(code = code, env = env, rebuild = rebuild, showOutput = showOutput,  : 
>   Error 1 occurred building shared library.

I am using Imager 0.2 and Rcpp 0.12.1.

Thanks!,
Michel

**Comment by dahtah** [comment]
Hi Michel. Sounds like a known clash between R.h and windows.h. Could you try this please?
cppFunction(foo.inline,depends="imager",includes="#define STRICT_R_HEADERS")

**Comment by mconsidine** [comment]
I am having a problem as well, using Rcpp 12.5, imager 0.2 and R 3-3.3.0
Here's what I get for an error message after the cppFunction call of above :+1: 
`> cppFunction(foo.inline,depends='imager',includes="#define STRICT_R_HEADERS")
H:/MattC/Work/Rtools/mingw_32/bin/g++  -I"C:/PROGRA~1/R/R-33~1.0/include" -DNDEBUG     -I"H:/MattC/Work/R/win-library/Rcpp/include" -I"H:/MattC/Work/R/win-library/imager/include" -I"C:/Users/MattC/AppData/Local/Temp/RtmpkDhNzm"  -I"d:/Compiler/gcc-4.9.3/local330/include"     -O2 -Wall  -mtune=core2 -c file18842ad52f87.cpp -o file18842ad52f87.o
In file included from H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/combaseapi.h:155:0,
                 from H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/objbase.h:14,
                 from H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/ole2.h:17,
                 from H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/shlobj.h:85,
                 from H:/MattC/Work/R/win-library/imager/include/CImg.h:171,
                 from H:/MattC/Work/R/win-library/imager/include/imager.h:21,
                 from file18842ad52f87.cpp:3:
H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/objidlbase.h:864:18: error: macro "Realloc" requires 3 arguments, but only 2 given
         SIZE_T cb) = 0;
                  ^
In file included from H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/combaseapi.h:155:0,
                 from H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/objbase.h:14,
                 from H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/ole2.h:17,
                 from H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/shlobj.h:85,
                 from H:/MattC/Work/R/win-library/imager/include/CImg.h:171,
                 from H:/MattC/Work/R/win-library/imager/include/imager.h:21,
                 from file18842ad52f87.cpp:3:
H:/MattC/Work/Rtools/mingw_32/i686-w64-mingw32/include/objidlbase.h:864:22: error: 'Realloc' declared as a 'virtual' field
         SIZE_T cb) = 0;
                      ^
In file included from C:/PROGRA~1/R/R-33~1.0/include/R.h:75:0,
                 from H:/MattC/Work/R/win-library/imager/include/imager.h:10,
                 from file18842ad52f87.cpp:3:
C:/PROGRA~1/R/R-33~1.0/include/R_ext/RS.h:71:37: error: expected identifier before '(' token
 #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
                                     ^
C:/PROGRA~1/R/R-33~1.0/include/R_ext/RS.h:71:47: error: 'parameter' declared as function returning a function
 #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
                                               ^
C:/PROGRA~1/R/R-33~1.0/include/R_ext/RS.h:71:50: error: expected ')' before ',' token
 #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
                                                  ^
C:/PROGRA~1/R/R-33~1.0/include/R_ext/RS.h:71:37: error: expected identifier before '(' token
 #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
                                     ^
C:/PROGRA~1/R/R-33~1.0/include/R_ext/RS.h:71:47: error: 'parameter' declared as function returning a function
 #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
                                               ^
C:/PROGRA~1/R/R-33~1.0/include/R_ext/RS.h:71:50: error: expected ')' before ',' token
 #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
                                                  ^
In file included from H:/MattC/Work/R/win-library/imager/include/imager.h:21:0,
                 from file18842ad52f87.cpp:3:
H:/MattC/Work/R/win-library/imager/include/CImg.h: In function 'const char* cimg_library::cimg::strbuffersize(UINT_PTR)':
H:/MattC/Work/R/win-library/imager/include/CImg.h:56410:82: warning: format '%lu' expects argument of type 'long unsigned int', but argument 4 has type 'UINT_PTR {aka unsigned int}' [-Wformat=]
     if (size<1024LU) cimg_snprintf(res,res._width,"%lu byte%s",size,size>1?"s":"");
                                                                                  ^
In file included from H:/MattC/Work/R/win-library/imager/include/imager.h:23:0,
                 from file18842ad52f87.cpp:3:
H:/MattC/Work/R/win-library/imager/include/wrappers.h: In function 'SEXPREC* Rcpp::wrap(const T&) [with T = cimg_library::CImgList<double>; SEXP = SEXPREC*]':
H:/MattC/Work/R/win-library/imager/include/wrappers.h:39:22: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for (int i = 0;i < ilist.size(); i++)
                      ^
make: *** [file18842ad52f87.o] Error 1
Warning message:
running command 'make -f "C:/PROGRA~1/R/R-33~1.0/etc/i386/Makeconf" -f "C:/PROGRA~1/R/R-33~1.0/share/make/winshlib.mk" SHLIB_LDFLAGS='$(SHLIB_CXXLDFLAGS)' SHLIB_LD='$(SHLIB_CXXLD)' SHLIB="sourceCpp_3.dll" OBJECTS="file18842ad52f87.o"' had status 2 
Error in sourceCpp(code = code, env = env, rebuild = rebuild, showOutput = showOutput,  : 
  Error 1 occurred building shared library.
>`

Any thoughts on how to approach this?
Thank you.

**Comment by dahtah** [comment]
OK, you're both having the same problem with header clashes. My guess is that the solution I suggested above doesn't work because the "includes" directive is inserted too late. 
To clarify, I've never tested this particular feature on a Windows machine but there's no reason it shouldn't work since imager compiles fine on Windows. 
I've made a new branch that contains a fix:
``` devtools::install_github("dahtah/imager",ref="cxxplugin") ```
hoping that works.

**Comment by mconsidine** [comment]
Something is occurring to me.  Is the fftw3 library required to be installed?  I looked at the changes you made on Github to create the new branch and it looks like that is the case.  Maybe I missed it, but it is not stated as being required for Windows on the docs page, but now that I look back it does seem to be cited for OSX and Linux.  I'll try setting that up and see if anything changes as the new branch didn't change anything for me.

**Comment by dahtah** [comment]
Yes, I'm afraid fftw is needed on Windows as well but I thought it was part of Rtools. Here's a link to a binary, you won't need to compile anything: 
http://win-builder.r-project.org/tnT7kwf59EfF/

**Comment by mconsidine** [comment]
There is an R package called 'fftw' that installs it.  Just did that, but no change.  Any idea where imager expects to find the fftw DLL?  i.e. is there a specific folder I should drop it into?

**Comment by dahtah** [comment]
It's not just about the dll, it's also the header files that are needed. There's probably a fftw3.h file somewhere on your drive, but right now I think you're better off installing the binary package.

**Comment by dahtah** [comment]
You won't need fttw to use CImg itself (for most of the API anyways)

**Comment by mconsidine** [comment]
Thanks for the replies.  Yeah, imager/CImg has been working so far.  I did just try the FFT example though and which it does run, the correct image doesn't come back.  This line
`sqrt(ff$re^2+ff$im^2) %>% plot(main="Power spectrum")`
generates this error
`Error in ff$re : $ operator is invalid for atomic vectors`
But this
`sqrt(Re(ff)^2+Im(ff)^2) %>% plot(main="pwr")`
runs.
The resulting "check" is not the same, by a long shot.

In any case, I digress.  I'll keep trying to get the API working, as I really like this library.
Matt
PS  Using a new install of R 3.3, Rstudio, Rtools and running on Windows 8.  The R software is 32-bit but the OS is 64-bit.  Now I'm wondering if therein might be a problem ...

**Comment by mconsidine** [comment]
Okay, installed the binary you pointed me to via Rstudio and the FFT test looks like it's working like it's supposed to.  More testing to come ...

**Comment by mconsidine** [comment]
Still getting an error, but looks like it might be fixable
`c:/Rtools/mingw_32/bin/g++ -shared -s -static-libgcc -o sourceCpp_3.dll tmp.def filec382f696180.o -lX11 -Ld:/Compiler/gcc-4.9.3/local330/lib/i386 -Ld:/Compiler/gcc-4.9.3/local330/lib -LC:/R/R-33~1.0/bin/i386 -lR
C:/Rtools/mingw_32/bin/../lib/gcc/i686-w64-mingw32/4.9.3/../../../../i686-w64-mingw32/bin/ld.exe: cannot find -lX11
collect2.exe: error: ld returned 1 exit status
Error in sourceCpp(code = code, env = env, rebuild = rebuild, showOutput = showOutput,  : 
  Error occurred building shared library.`

Looks like this is a vestige of a Linux/Mac environment that uses X11?  Is there a file that I can edit to drop this call or substitute it with something more appropriate?

**Comment by mconsidine** [comment]
Looks like the source of the call to the X11 lib is 
`https://github.com/dahtah/imager/blob/cxxplugin/R/inline.R`
And it doesn't appear to be o/s dependent.  Has inline compilation been confirmed to worked on Windows by anyone?

**Comment by dahtah** [comment]
All right, let me try and rewrite the Rcpp inline plugin. In the meanwhile I guess a quick and dirty fix is to copy a random library and name the copy libX11.a (or whatever the right file extension is on Windows) so that the linker doesn't complain. 
To answer your question, no, I don't think anyone's used inline compilation yet on Windows that I know of. Windows support is a major headache.

**Comment by mconsidine** [comment]
If I take another DLL, rename it to X11.DLL and place it in the i386 bin folder for R, then I get an error about a number of undeclared references.  Googling those errors suggests that other's solutions using CImg was to link against libgdi32 on Windows :
  http://stackoverflow.com/questions/7861517/how-to-draw-a-simple-function-using-the-cimg-library

Hope that helps a bit (though I doubt it if you don't use Windows :) )

**Comment by mconsidine** [comment]
I think I have a workaround!  I found that libgdi32.a is in Rtools, specifically Rtools/bin/mingw_32/i686-w64-mingw32/lib (version appropriate for R 3-3.3.0).  So I took that and copied it to R/R-3.3.0/bin/i386 and renamed it to X11.dll.

At this point I ran the code found here 
  http://dahtah.github.io/imager/cpp_api.html
and the cppFunction now runs without error.  I did *NOT* need to put the STRICT HEADERS bit in.

The properly blurred and erode birds shows up just fine.

Hope this helps!
Matt

**Comment by mconsidine** [comment]
I can confirm that this works on a completely different Windows installation.  Using R-3.3.0 and the latest Rtools.  This trick doesn't work with the CRAN version of imager but with the win-builder version linked to earlier in this thread.  With that, the call to cppFunction does not need 
`includes="#define STRICT_R_HEADERS"`
added.

Regards,
Matt

**Comment by dahtah** [comment]
Glad it works. I've changed the way the inline plugin is defined, so that it now works without the hack (a least on the Windows box at winbuilder.org). 
New binary at: http://win-builder.r-project.org/X65gRLb6w2Dr

**Comment by mconsidine** [comment]
Thanks for the new binary.  I've loaded this up and can confirm it works without having to use either the STRICT_R_HEADERS parameter to cppFunction  or the renamed library trick.
Much obliged!
Matt
(PS and unrelated : How would I subtract one blur from another in the inline code?  I've tried img=img.blur(1)-img.blur(3) and variants thereof but that's not working.  Is there a compact/elegant way I am missing?)
